### ISSUE TITLE
MicroBitUARTService.read(buf,len,ASYNC) looses data when len == 1

### ISSUE THREAD

**Issue by drmpf** [lancaster-university/microbit-dal#201]
This small sample BLE Uart program looses data.
Send {.}   displays {} on the terminal I.e. the second byte  ,  . (dot),  is missing 
change the len to 22 and then send 
{.} and get {.} displayed on the terminal

```
#include "MicroBit.h"
#include "MicroBitUARTService.h"
MicroBit uBit;
MicroBitUARTService *uart;

int connected = 1;
void onConnected(MicroBitEvent) {
    connected = 1;
    uBit.serial.send("Connected\n");
}

void processBLEUart() {
    uint8_t readBuf[22];
    while (1) { // loop for ever
        if ((!connected) || (!uart->isReadable())) {
            uBit.sleep(50); // wait 50mS and check again
            continue; // loop;
        }
        //else we are connected AND there is some input to read
        if (uart->isReadable()) {
            while (uart->isReadable()) {
                int charCount = uart->read(readBuf, 1, ASYNC); // looses second byte send, setting len == 22 gets all three
                if (charCount != 0) {
                    for (int i = 0; i < charCount; i++)
                        uBit.serial.send((char) readBuf[i]);
                }
            }
            uBit.serial.send('\n');
        }
    }
}

void onDisconnected(MicroBitEvent) {
    uBit.serial.send("Disconnected\n");
    connected = 0;
}

int main() {
    // Initialise the micro:bit runtime.
    uBit.init();

    uBit.messageBus.listen(MICROBIT_ID_BLE, MICROBIT_BLE_EVT_CONNECTED, onConnected);
    uBit.messageBus.listen(MICROBIT_ID_BLE, MICROBIT_BLE_EVT_DISCONNECTED, onDisconnected);

    uart = new MicroBitUARTService(*uBit.ble, 32, 32);
    uBit.serial.send("UART ready\n");
    create_fiber(&processBLEUart); // create fiber and schedule it.
    release_fiber();
}

```

**Comment by finneyj** [comment]
@drmpf 

weird... The code is very straightforward here - just a ring buffer. Not at all obvious what could be causing this behaviour.

Can you please tray a few additional tests to help isolate the problem? Suggest:

1) Sending a large set of bytes and see if there's a pattern? Say send "abcdefgh" (i.e. is it somehow dropping every other byte?).

2) Just to eliminate any possibility of this being an issue in the output serial (unlikely), maintain a count of bytes received in your code and display that at the end, and check consistency against what is printed...

**Comment by cpseager** [comment]
Do you have a race where uart->read is updating readBuf before the previous uBit.serial.send has finished using the old readBuf?

**Comment by drmpf** [comment]
2) This modified method always prints 2 for the total chars.  When chars sent are {.} and {!}

```
void processBLEUart() {
    uint8_t readBuf[22];
    while (1) { // loop for ever
        if ((!connected) || (!uart->isReadable())) {
            uBit.sleep(50); // wait 50mS and check again
            continue; // loop;
        }
        //else we are connected AND there is some input to read
        if (uart->isReadable()) {
            int totalCount = 0;
            while (uart->isReadable()) {
                int charCount = uart->read(readBuf, 1, ASYNC); // looses second byte send, setting len == 22 gets all three
                if (charCount != 0) {
                    totalCount += charCount;
                    for (int i = 0; i < charCount; i++) {
                       // uBit.serial.send((char) readBuf[i]);
                    }
                }
            }
            uBit.serial.send(totalCount);            
            uBit.serial.send('\n');
        }
    }
}

```

**Comment by jamesadevine** [comment]
@drmpf What characters are you sending normally?

**Comment by drmpf** [comment]
for now I am just sending {.} and {!}  more later
but .. doing you suggested tests

1)
sending  with len == 22 output is 
1234567890abcdef{.}19   
where 19 is the count

setting len == 1 gives as output
13579ace{}10
i.e. every second char is missing

```
void processBLEUart() {
    uint8_t readBuf[22];
    while (1) { // loop for ever
        if ((!connected) || (!uart->isReadable())) {
            uBit.sleep(50); // wait 50mS and check again
            continue; // loop;
        }
        //else we are connected AND there is some input to read
        if (uart->isReadable()) {
            int totalCount = 0;
            while (uart->isReadable()) {
                int charCount = uart->read(readBuf, 1, ASYNC); // looses second byte send, setting len == 22 gets all three
                if (charCount != 0) {
                    totalCount += charCount;
                    for (int i = 0; i < charCount; i++) {
                       uBit.serial.send((char) readBuf[i]);
                    }
                }
            }
            uBit.serial.send(totalCount);            
            uBit.serial.send('\n');
        }
    }
}

```

2) commenting out the 
uBit.serial.send((char) readBuf[i]);  line
gives output
10
as does commenting out the whole 

 ```
 for (int i = 0; i < charCount; i++) {
                       uBit.serial.send((char) readBuf[i]);
                    }
```

**Comment by finneyj** [comment]
@cpseager Good idea... but I'm pretty sure that serial send() operation will put it in a tx buffer, or block the fiber if the tx buffer is full: 

https://github.com/lancaster-university/microbit-dal/blob/master/source/drivers/MicroBitSerial.cpp#L413

**Comment by drmpf** [comment]
removed the serial.send inside the loop, see above tests, same result.

**Comment by finneyj** [comment]
@drmpf thanks. This confirms its in the receive buffer handling of the UART service somewhere.  hmm....

**Comment by finneyj** [comment]
This looks a little suspicious: 

https://github.com/lancaster-university/microbit-dal/blob/master/source/bluetooth/MicroBitUARTService.cpp#L359

Won't this evaluate the read operation *before* checking that the buffer is full? Flipping the order of the terms might be good idea...

p.s. Wonder also if this test should be >=0 as '0' is a valid character in a bytestream? @jamesadevine?

**Comment by jamesadevine** [comment]
Yeah that doesn't look right, should definitely be >= 0

**Comment by finneyj** [comment]
cool. @drmpf - can you try this simple patch on your build and test for us?

Just update this line: https://github.com/lancaster-university/microbit-dal/blob/master/source/bluetooth/MicroBitUARTService.cpp#L359

to something like:

```
while(i < len && (c = getc(mode)) > 0)
```

**Comment by drmpf** [comment]
Ok changed the code to
      `  while( (i<len) && ((c = getc(mode)) >= 0) )`
and that fixed the problem

Also
currently MICROBIT_NO_DATA  = -1012  in ErrorNo.h
but
`char c = (char) uart->getc(); `   when there is no data
returns 0x0C at the moment, which is valid utf-8 byte.
suggest changing 
MICROBIT_NO_DATA  to  -1
so that 
` (char) MICROBIT_NO_DATA`  or `(uint8_t) MICROBIT_NO_DATA`
   becomes 0xff  which is not a valid utf-8 byte.
AND change the above test to
   `  while( (i<len) && ((c = getc(mode)) != -1) )`
for clarity

**Comment by jamesadevine** [comment]
@drmpf That api returns an int purely for that reason. 0xff can also be a valid byte.

**Comment by drmpf** [comment]
Assuming you are supporting UTF-8 for non-English speaking users and for symbols like degC,  I disagree,  https://en.wikipedia.org/wiki/UTF-8  shows the valid UTF-8 bytes. 
  0xff is not included in the table.
In UTF-8, "Code points larger than 127 _(i.e. 0xff)_  are represented by multi-byte sequences" none of the bytes in these multi-byte sequences includes the byte 0xff

While the api returns an int, in almost all cases the result will be stored in a char or uint8_t array.  Given that 0xff is not a valid UTF-8 byte, storing -1 in a char or uint8_t array preserves the no data flag, which can be detected later, while -1012 will be stored as 0x0C and so looks like valid data in the array.

**Comment by DavidWhaleIET** [comment]
What about binary data packets that are not strings? 0xff is valid in those?

**Comment by finneyj** [comment]
Agree - it's important that plain old unsigned bytes are supported through this API. Testing the return code is possible with the API (as @jamesadevine says - that's why it returns a data type with more than 8 bits). 

Perhaps we should tighten up the example in the docs to provide an example where showing that testing should come before a cast?

**Comment by drmpf** [comment]
I agree the return needs to be checked against an 'out-of-band' constant.
BUT
any quick scan of C/C++ docs for getc() will show it returns EOF on error and that EOF is generally -1.  
The current return of -1012, will be un-expected by users with any C/C++ coding background.

 In my experience most C/C++ programmer ignore error returns until the program breaks.  Erroneous 0xff bytes/chars are more easily recognized as an error than extraneous form feeds, 0x0C, appearing in the output.

**Comment by finneyj** [comment]
It is good to maintain conventions where possible - and I totally agree that most C programmers only check error codes when debugging. :-)

This error code MICROBIT_NO_DATA is used in a few places across microbit-dal, so I'd be slightly reluctant to change it, as it could break peoples programs (folks *should* compare return codes to labels rather than literal integers, but they don't always do that... especially folks that are less experienced). There's also then a ripple effect - if we change this one error code, why not the others, and what would we change them all to - they can't all be the same...

Unless there's a lot of support for changing error code values, I don't think we should change them.

**Comment by finneyj** [comment]
@drmpf - Thanks for all the testing! As you have the working test case, would you like to submit a short Pull Request with the fix?
